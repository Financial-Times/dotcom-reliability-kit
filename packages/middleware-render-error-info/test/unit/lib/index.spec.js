const createErrorRenderingMiddleware = require('../../../lib/index');

jest.mock('@dotcom-reliability-kit/serialize-error', () => jest.fn());
const serializeError = require('@dotcom-reliability-kit/serialize-error');
serializeError.mockReturnValue({
	// We nest multiple causes so that we can test each
	// kind of error warning which might appear in the
	// output. E.g. operational error vs non-error thrown
	cause: {
		cause: {
			data: {}
		},
		data: {},
		isOperational: false,
		message: 'mock caused error message',
		relatesToSystems: [],
		stack: 'mock caused error stack'
	},
	code: 'MOCK_SERIALIZED_ERROR_CODE',
	data: {
		booleanValue: true,
		stringValue: 'mock value',
		numberValue: 123,
		arrayValue: ['mock', 'array'],
		objectValue: { isObject: true }
	},
	isOperational: true,
	message: 'mock serialized error message',
	name: 'MockSerializedError',
	relatesToSystems: ['mock-system-1', 'mock-system-2'],
	stack: 'mock serialized error stack <script>oops</script>',
	statusCode: 456
});

jest.mock('@dotcom-reliability-kit/log-error', () => ({
	logRecoverableError: jest.fn()
}));
const { logRecoverableError } = require('@dotcom-reliability-kit/log-error');

jest.mock('@dotcom-reliability-kit/app-info', () => ({}));
const appInfo = require('@dotcom-reliability-kit/app-info');

describe('@dotcom-reliability-kit/middleware-render-error-info', () => {
	let middleware;

	beforeEach(() => {
		appInfo.environment = 'development';
		appInfo.systemCode = 'mock-system-code';
		middleware = createErrorRenderingMiddleware();
	});

	describe('.default', () => {
		it('aliases the module exports', () => {
			expect(createErrorRenderingMiddleware.default).toStrictEqual(
				createErrorRenderingMiddleware
			);
		});
	});

	describe('middleware(error, request, response, next)', () => {
		let error;
		let next;
		let request;
		let response;

		beforeEach(() => {
			// The error page template contains the current year
			jest.useFakeTimers().setSystemTime(new Date('1888-01-09'));

			error = new Error('mock error');
			request = {
				method: 'mock request method',
				path: 'mock request path',
				query: {
					'mock-request-query': 'yes',
					'x-api-key': 'foo' // 'x-api-key' is a concealed request property name
				},
				headers: {
					'mock-request-header': 'yes',
					cookie: 'bar' // 'cookie' is a concealed request property name
				}
			};
			response = {
				send: jest.fn(),
				set: jest.fn(),
				status: jest.fn()
			};
			next = jest.fn();

			middleware(error, request, response, next);
		});

		it('serializes the error', () => {
			expect(serializeError).toBeCalledWith(error);
		});

		it('responds with the serialized error status code', () => {
			expect(response.status).toBeCalledTimes(1);
			expect(response.status).toBeCalledWith(456);
		});

		it('responds with a Content-Type header of "text/html"', () => {
			expect(response.set).toBeCalledTimes(1);
			expect(response.set).toBeCalledWith('content-type', 'text/html');
		});

		it('responds with an HTML representation of the error', () => {
			expect(response.send).toBeCalledTimes(1);
			const html = response.send.mock.calls[0][0];
			expect(typeof html).toBe('string');
			// We replace multiple line break characters with a single one, and
			// remove all tab characters so that formatting whitespace changes
			// in the markup don't break the tests
			expect(
				html
					.replace(/[\r\n]+/g, '\n')
					.replace(/\t+/g, '')
					.trim()
			).toMatchSnapshot();
		});

		it('does not call `next` with the original error', () => {
			expect(next).toBeCalledTimes(0);
		});

		describe('when the system code is not set', () => {
			beforeEach(() => {
				appInfo.systemCode = null;
				middleware = createErrorRenderingMiddleware();
				response.send = jest.fn();
				middleware(error, request, response, next);
			});

			it('uses the default system code in the Origami Build Service URLs', () => {
				expect(response.send).toBeCalledTimes(1);
				const html = response.send.mock.calls[0][0];
				expect(typeof html).toBe('string');

				// Regular expressions expecting the URL-encoded Origami default system code
				expect(html).toMatch(
					/bundles\/css\?system_code=%24%24%24-no-bizops-system-code-%24%24%24/i
				);
				expect(html).toMatch(
					/bundles\/js\?system_code=%24%24%24-no-bizops-system-code-%24%24%24/i
				);
			});

			it('uses a default "application" string in the page title', () => {
				expect(response.send).toBeCalledTimes(1);
				const html = response.send.mock.calls[0][0];
				expect(typeof html).toBe('string');
				expect(html).toMatch(
					/<title>MockSerializedError in application<\/title>/
				);
			});
		});

		describe('when the serialized error does not have a `statusCode` property', () => {
			beforeEach(() => {
				serializeError.mockReturnValue({
					statusCode: null,
					data: {}
				});
				response.status = jest.fn();

				middleware(error, request, response, next);
			});

			it('responds with a 500 status code', () => {
				expect(response.status).toBeCalledTimes(1);
				expect(response.status).toBeCalledWith(500);
			});
		});

		describe('when the environment is set to "production"', () => {
			beforeEach(() => {
				appInfo.environment = 'production';
				middleware = createErrorRenderingMiddleware();
				response = {
					send: jest.fn(),
					set: jest.fn(),
					status: jest.fn()
				};
				next = jest.fn();
				middleware(error, request, response, next);
			});

			it('does not render and send an error info page', () => {
				expect(response.status).toBeCalledTimes(0);
				expect(response.set).toBeCalledTimes(0);
				expect(response.send).toBeCalledTimes(0);
			});

			it('calls `next` with the original error', () => {
				expect(next).toBeCalledWith(error);
			});
		});

		describe('when rendering the page fails', () => {
			let renderingError;

			beforeEach(() => {
				renderingError = new Error('rendering failed');
				response.send.mockImplementation(() => {
					throw renderingError;
				});
				next = jest.fn();
				middleware(error, request, response, next);
			});

			it('logs the rendering error as recoverable', () => {
				expect(logRecoverableError).toBeCalledTimes(1);
				expect(logRecoverableError).toBeCalledWith({
					error: renderingError,
					request
				});
			});

			it('calls `next` with the original error', () => {
				expect(next).toBeCalledWith(error);
			});
		});
	});
});
