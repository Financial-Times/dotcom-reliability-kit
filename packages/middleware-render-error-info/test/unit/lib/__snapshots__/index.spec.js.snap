// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`@dotcom-reliability-kit/middleware-render-error-info middleware(error, request, response, next) responds with an HTML representation of the error 1`] = `
"<!DOCTYPE html>
<html lang="en">
<head>
<meta charset="utf-8" />
<meta http-equiv="X-UA-Compatible" content="IE=Edge" />
<title>MockSerializedError in mock-system-code</title>
<meta name="viewport" content="initial-scale=1.0, width=device-width" />
<link rel="stylesheet" type="text/css" href="https://www.ft.com/__origami/service/build/v3/bundles/css?system_code=mock-system-code&brand=internal&components=o-autoinit%40%5E3.1.3%2Co-spacing%40%5E3.2.1%2Co-colors%40%5E6.4.2%2Co-layout%40%5E5.2.3%2Co-header-services%40%5E5.2.3%2Co-footer-services%40%5E4.2.3%2Co-syntax-highlight%40%5E4.2.3%2Co-message%40%5E5.3.1" />
<style>
/* Override to avoid super long pages and make pre elements scroll */
pre {
max-height: 300px;
white-space: pre-wrap;
}
/* Override to add some spacing below alert messages in the main body of the page */
.o-layout__main .o-message {
margin-bottom: var(--o-spacing-s4);
}
/* Override to add some spacing below sections in the main body of the page */
.o-layout__main > section {
margin-bottom: var(--o-spacing-m12);
}
/* Styles for a key/value list. Description list styles are not provided by Origami */
.kv-list {
display: grid;
grid-template-columns: 1fr 4fr;
margin: 0;
}
.kv-list__key {
padding-right: var(--o-spacing-s3);
}
.kv-list__label {
font-weight: 500;
}
.kv-list__value {
overflow: auto;
background-color: var(--o-colors-black-5);
padding: var(--o-spacing-s2);
}
.kv-list__key,
.kv-list__value {
margin: 0;
padding-top: var(--o-spacing-s2);
padding-bottom: var(--o-spacing-s2);
border-bottom: 1px solid var(--o-colors-black-20);
}
.kv-list__key:first-child,
.kv-list__value:nth-child(2) {
border-top: 1px solid var(--o-colors-black-20);
}
.kv-list__value pre {
margin: 0;
padding: 0;
}
/* Boolean value styling */
.boolean {
font-weight: 600;
}
.boolean--positive {
color: var(--o-colors-jade);
}
.boolean--negative {
color: var(--o-colors-crimson);
}
</style>
<script defer="" src="https://www.ft.com/__origami/service/build/v3/bundles/js?system_code=mock-system-code&brand=internal&components=o-autoinit%40%5E3.1.3%2Co-spacing%40%5E3.2.1%2Co-colors%40%5E6.4.2%2Co-layout%40%5E5.2.3%2Co-header-services%40%5E5.2.3%2Co-footer-services%40%5E4.2.3%2Co-syntax-highlight%40%5E4.2.3%2Co-message%40%5E5.3.1"></script>
</head>
<body>
<div class="o-layout o-layout--docs" data-o-component="o-layout" data-o-layout-construct-nav="true">
<div class="o-layout__header">
<header class="o-header-services" data-o-component="o-header-services">
<div class="o-header-services__top">
<div class="o-header-services__logo"></div>
<div class="o-header-services__title">
<a class="o-header-services__product-name" href="undefined">
Reliability Kit Error Info Page
</a>
</div>
</div>
</header>
<div class="o-message o-message--alert o-message--error" data-o-component="o-message" data-o-message-close="false">
<div class="o-message__container">
<div class="o-message__content">
<p class="o-message__content-main">
<span class="o-message__content-highlight">Welcome to the Reliability Kit error info page.</span><br/>
This page has appeared because you're using the
<a href="https://github.com/Financial-Times/dotcom-reliability-kit/tree/main/packages/middleware-render-error-info#readme">@dotcom-reliability-kit/middleware-render-error-info</a>
package in your app.<br/>
Hopefully this will make it a lot easier for you to debug the cause of this error.<br/>
Don't panic: this only appears when you're running the app locally!
</p>
</div>
</div>
</div>
</div>
<div class="o-layout__sidebar o-layout-typography"></div>
<main class="o-layout__main o-layout-typography" data-o-component="o-syntax-highlight">
<h1 id="errors">Error information</h1>

<section class="o-layout__main__full-span">
<h2 id="error-MOCK_SERIALIZED_ERROR_CODE-mock-serialized-error-message">MockSerializedError</h2>

<dl class="kv-list">

<dt class="kv-list__key">
<span class="kv-list__label">Message:</span>
<br/><small>The human-readable message which describes this error</small>
</dt>
<dd class="kv-list__value">mock serialized error message</dd>

<dt class="kv-list__key">
<span class="kv-list__label">Code:</span>
<br/><small>The machine-readable code which identifies this error</small>
</dt>
<dd class="kv-list__value"><pre><code>MOCK_SERIALIZED_ERROR_CODE</code></pre></dd>

<dt class="kv-list__key">
<span class="kv-list__label">Status Code:</span>
<br/><small>The HTTP status code this error should be served under</small>
</dt>
<dd class="kv-list__value"><pre><code>456</code></pre></dd>

<dt class="kv-list__key">
<span class="kv-list__label">Is operational:</span>
<br/><small>Whether this is a known or expected error (<a href="https://github.com/Financial-Times/dotcom-reliability-kit/blob/main/docs/getting-started/throwing-errors.md#operational-errors" target="_blank">docs</a>)</small>
</dt>
<dd class="kv-list__value"><span class="boolean boolean--positive">Yes</span></dd>

<dt class="kv-list__key">
<span class="kv-list__label">Related systems:</span>
<br/><small>Any Biz Ops systems which this error is related to</small>
</dt>
<dd class="kv-list__value">
<a
href="https://biz-ops.in.ft.com/System/mock-system-1"
target="_blank"
>mock-system-1</a>
, 
<a
href="https://biz-ops.in.ft.com/System/mock-system-2"
target="_blank"
>mock-system-2</a>
</dd>

<dt class="kv-list__key">
<span class="kv-list__label">Data (booleanValue):</span>

</dt>
<dd class="kv-list__value"><pre><code>true</code></pre></dd>

<dt class="kv-list__key">
<span class="kv-list__label">Data (stringValue):</span>

</dt>
<dd class="kv-list__value"><pre><code>&quot;mock value&quot;</code></pre></dd>

<dt class="kv-list__key">
<span class="kv-list__label">Data (numberValue):</span>

</dt>
<dd class="kv-list__value"><pre><code>123</code></pre></dd>

<dt class="kv-list__key">
<span class="kv-list__label">Data (arrayValue):</span>

</dt>
<dd class="kv-list__value"><pre><code>[
    &quot;mock&quot;,
    &quot;array&quot;
]</code></pre></dd>

<dt class="kv-list__key">
<span class="kv-list__label">Data (objectValue):</span>

</dt>
<dd class="kv-list__value"><pre><code>{
    &quot;isObject&quot;: true
}</code></pre></dd>

<dt class="kv-list__key">
<span class="kv-list__label">Fingerprint:</span>
<br/><small>A unique hash generated from this error's stack</small>
</dt>
<dd class="kv-list__value"><pre><code>mock serialized error fingerprint</code></pre></dd>

<dt class="kv-list__key">
<span class="kv-list__label">Stack:</span>
<br/><small>The full stack trace for the error, indicating where it was thrown</small>
</dt>
<dd class="kv-list__value"><pre><code>mock serialized error stack &lt;script&gt;oops&lt;/script&gt;</code></pre></dd>

<dt class="kv-list__key">
<span class="kv-list__label">Cause:</span>
<br/><small>The original thrown error which resulted in this operational error</small>
</dt>
<dd class="kv-list__value">
<section class="o-layout__main__full-span">
<h2 id="error-undefined-mock-caused-error-message">undefined</h2>

<div class="o-message o-message--notice o-message--warning-light" data-o-component="o-message" data-o-message-close="false">
<div class="o-message__container o-layout__unstyled-element">
<div class="o-message__content o-layout__unstyled-element">
<p class="o-message__content-main o-layout__unstyled-element">
<span class="o-message__content-highlight o-layout__unstyled-element">This error is unexpected.</span><br/>
It is not extending the <a href="https://github.com/Financial-Times/dotcom-reliability-kit/tree/main/packages/errors#operationalerror" target="_blank">OperationalError</a>
class provided by Reliability Kit which indicates to us that we don't understand how to manage this error.
If this is not true, then consider catching this error and throwing an operational error.
See the <a href="https://github.com/Financial-Times/dotcom-reliability-kit/blob/main/docs/getting-started/throwing-errors.md" target="_blank">throwing errors guide</a> for more information.

</p>
</div>
</div>
</div>

<dl class="kv-list">

<dt class="kv-list__key">
<span class="kv-list__label">Message:</span>
<br/><small>The human-readable message which describes this error</small>
</dt>
<dd class="kv-list__value">mock caused error message</dd>

<dt class="kv-list__key">
<span class="kv-list__label">Is operational:</span>
<br/><small>Whether this is a known or expected error (<a href="https://github.com/Financial-Times/dotcom-reliability-kit/blob/main/docs/getting-started/throwing-errors.md#operational-errors" target="_blank">docs</a>)</small>
</dt>
<dd class="kv-list__value"><span class="boolean boolean--negative">No</span></dd>

<dt class="kv-list__key">
<span class="kv-list__label">Related systems:</span>
<br/><small>Any Biz Ops systems which this error is related to</small>
</dt>
<dd class="kv-list__value">None</dd>

<dt class="kv-list__key">
<span class="kv-list__label">Stack:</span>
<br/><small>The full stack trace for the error, indicating where it was thrown</small>
</dt>
<dd class="kv-list__value"><pre><code>mock caused error stack</code></pre></dd>

<dt class="kv-list__key">
<span class="kv-list__label">Cause:</span>
<br/><small>The original thrown error which resulted in this operational error</small>
</dt>
<dd class="kv-list__value">
<section class="o-layout__main__full-span">
<h2 id="error-undefined-undefined">undefined</h2>

<div class="o-message o-message--notice o-message--warning-light" data-o-component="o-message" data-o-message-close="false">
<div class="o-message__container o-layout__unstyled-element">
<div class="o-message__content o-layout__unstyled-element">
<p class="o-message__content-main o-layout__unstyled-element">
<span class="o-message__content-highlight o-layout__unstyled-element">A non-Error object was thrown.</span><br/>
The thing that was thrown does not extend the Node.js built-in Error class.
This makes it much harder to debug the issue, as we cannot find the line numbers where
the error occurred. You'll need to search your source code for anywhere that a non-Error
appears after the "throw" keyword.
See the <a href="https://github.com/Financial-Times/dotcom-reliability-kit/blob/main/docs/getting-started/throwing-errors.md#using-error-objects" target="_blank">using error objects guide</a> for more information.

</p>
</div>
</div>
</div>

<dl class="kv-list">

</dl>
</section>
</dd>

</dl>
</section>
</dd>

</dl>
</section>


<section class="o-layout__main__full-span">
<h2 id="request">HTTP Request</h2>

<dl class="kv-list">

<dt class="kv-list__key">
<span class="kv-list__label">Method:</span>

</dt>
<dd class="kv-list__value"><pre><code>mock request method</code></pre></dd>

<dt class="kv-list__key">
<span class="kv-list__label">Path:</span>

</dt>
<dd class="kv-list__value"><pre><code>mock request path</code></pre></dd>

<dt class="kv-list__key">
<span class="kv-list__label">Query (mock-request-query):</span>

</dt>
<dd class="kv-list__value"><pre><code>yes</code></pre></dd>

<dt class="kv-list__key">
<span class="kv-list__label">Query (x-api-key):</span>

</dt>
<dd class="kv-list__value"><pre><code>** This value is likely sensitive and should be examined via other tools **</code></pre></dd>

<dt class="kv-list__key">
<span class="kv-list__label">Header (mock-request-header):</span>

</dt>
<dd class="kv-list__value"><pre><code>yes</code></pre></dd>

<dt class="kv-list__key">
<span class="kv-list__label">Header (cookie):</span>

</dt>
<dd class="kv-list__value"><pre><code>** This value is likely sensitive and should be examined via other tools **</code></pre></dd>

</dl>
</section>


<section class="o-layout__main__full-span">
<h2 id="response">HTTP Response</h2>

<dl class="kv-list">

</dl>
</section>


<section class="o-layout__main__full-span">
<h2 id="app">App Info</h2>

<dl class="kv-list">

<dt class="kv-list__key">
<span class="kv-list__label">System:</span>
<br/><small>The system this error is being served by</small>
</dt>
<dd class="kv-list__value">
<a
href="https://biz-ops.in.ft.com/System/mock-system-code"
target="_blank"
>mock-system-code</a>
</dd>

<dt class="kv-list__key">
<span class="kv-list__label">Environment:</span>
<br/><small>The environment (production/development) that the app is running in</small>
</dt>
<dd class="kv-list__value">development</dd>

</dl>
</section>

</main>
<div class="o-layout__footer">
<footer class="o-footer-services">
<div class="o-footer-services__container">
<div class="o-footer-services__wrapper o-footer-services__wrapper--top">
<a class="o-footer-services__icon-link o-footer-services__icon-link--github" href="https://github.com/Financial-Times/dotcom-reliability-kit#readme">Reliability Kit Documentation</a>
<a class="o-footer-services__icon-link o-footer-services__icon-link--slack" href="https://financialtimes.slack.com/archives/C02B89GEQHF">#dotcom-support</a>
</div>
</div>
<div class="o-footer-services__container">
<div class="o-footer-services__wrapper o-footer-services__wrapper--legal">
<div class="o-footer-services__links">
<a href="http://help.ft.com/help/legal-privacy/cookies/">Cookies</a>
<a href="http://help.ft.com/help/legal-privacy/copyright/copyright-policy/">Copyright</a>
<a href="http://help.ft.com/help/legal-privacy/privacy/">Privacy</a>
<a href="http://help.ft.com/help/legal-privacy/terms-conditions">Terms &amp; Conditions</a>
</div>
<p>
<span>&#xA9; THE FINANCIAL TIMES LTD 1888.</span> FT and
&apos;Financial Times&apos; are trademarks of The Financial Times Ltd.
</p>
</div>
</div>
</footer>
</div>
</div>
</body>
</html>"
`;
